### PR TITLE
Updating the LBANN, Hydrogen, and DiHydrogen recipes

### PR DESCRIPTION
**Pull request by bvanessen**
Updating the LBANN, Hydrogen, and DiHydrogen recipes for both new
variants and to make sure that RPATHs are properly setup.

### CODE REVIEW

**Comment by bvanessen** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/hydrogen/package.py [hunk 1]
```diff
@@ -90,6 +90,7 @@ class Hydrogen(CachedCMakePackage, CudaPackage, ROCmPackage):
     # Note that this forces us to use OpenBLAS until #1712 is fixed
     depends_on("openblas", when="blas=openblas")
     depends_on("openblas +ilp64", when="blas=openblas +int64_blas")
+    depends_on("openblas@0.3.21:0.3.23", when="blas=openblas arch=ppc64le:")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/hydrogen/package.py:93]: (not a request, just a question) The `blas` variant is a bit weird. Is there any reason why this is / was necessary?

File: var/spack/repos/builtin/packages/lbann/package.py [hunk 2]
```diff
@@ -272,15 +287,36 @@ def initconfig_compiler_entries(self):
             # of CMake
             entries.append(cmake_cache_option("CMAKE_EXPORT_COMPILE_COMMANDS", True))
 
+        entries.append(cmake_cache_string("CMAKE_INSTALL_RPATH_USE_LINK_PATH", "ON"))
+        linker_flags = "-Wl,--disable-new-dtags"
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/lbann/package.py:291]: (Only blocking question) Will this setting conflict with `config:shared_linking:type:runpath`?

**Review by alalazo — COMMENTED** [review]
I have one question, but otherwise LGTM.

File: var/spack/repos/builtin/packages/lbann/package.py (hunk 2 above)
Comment (bvanessen) [reply var/spack/repos/builtin/packages/lbann/package.py:291]: @alalazo this flag `entries.append(cmake_cache_string("CMAKE_INSTALL_RPATH_USE_LINK_PATH", "ON"))` matches what is in the `cmake` package.  The  `linker_flags = "-Wl,--disable-new-dtags"` is used to make sure that `RPATH` works properly on all of the systems that we are working with.

File: var/spack/repos/builtin/packages/hydrogen/package.py (hunk 1 above)
Comment (bvanessen) [reply var/spack/repos/builtin/packages/hydrogen/package.py:93]: @alalazo We find that many versions of `openblas` fail to build on power architectures.  This is to help preserve sanity.

File: var/spack/repos/builtin/packages/lbann/package.py (hunk 2 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/lbann/package.py:291]: It's just the linker flag I'm concerned about. It's a bad idea to mix RPATH and RUNPATHs, and if a user sets `config:shared_linking:type:runpath` there will be RUNPATH set on dependencies. Do you know why the flag was necessary? That flag is already supposed to be injected in our compiler wrapper based on the config value - so it shouldn't be necessary.

File: var/spack/repos/builtin/packages/hydrogen/package.py (hunk 1 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/hydrogen/package.py:93]: @bvanessen Fine, this was just a curiosity - but I was asking about the pre-existing `blas=XXX` variant, not about the version range :thinking:

**Review by alalazo — APPROVED** [review]
@bvanessen Merging this as is, but we should figure out why this flag is needed at all

**Pull request merged**